python/pyspark/sql/types.py

```diff
@@ -191,14 +191,25 @@ def needConversion(self):
 
     def toInternal(self, dt):
         if dt is not None:
-            seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
-                       else time.mktime(dt.timetuple()))
+            seconds = 0.0
+            try:
+                seconds = (calendar.timegm(dt.utctimetuple()) if dt.tzinfo
+                           else time.mktime(dt.timetuple()))
+            except:
```

Comment:
I think we shouldn't better rely on exception handling for regular data parsing path.

Reply:
Can we do this with an if-else with OS and negative value check?

Reply:
Sure, I'll change the test again.